src/common/marlin_server.cpp

```diff
@@ -581,6 +589,61 @@ static void _server_print_loop(void) {
     default:
         break;
     }
+    if (fan_check_enable) {
+        fan_time++;
+
+        if (fan_time >= scan_fan_time) {
+            if (fanctl1.getActualRPM() == 0 && marlin_server.vars.temp_nozzle > EXTRUDER_AUTO_FAN_TEMPERATURE + 5
+                && error_dialog_show == false) {
+                fan_time = 0;
+                _dbg("nozzle fan not spin!");
+                nozzle_fan_error++;
+            } else if (fanctl0.getActualRPM() == 0 && fanctl0.getPWM() > 0 && error_dialog_show == false) {
+                fan_time = 0;
+                _dbg("print fan not spin!");
+                print_fan_error++;
+            } else if (fan_time >= 5000) {
+                if (fanctl1.getActualRPM() > minimal_fan_rpm_for_detect_spining) {
+                    error_dialog_show = false;
+                } else if (fanctl0.getActualRPM() > minimal_fan_rpm_for_detect_spining && fanctl0.getPWM() > 0) {
+                    error_dialog_show = false;
+                }
+                fan_time = 0;
+            }
+        }
+
+        if (nozzle_fan_error == 5) {
+            set_warning(WarningType::HotendFanError);
+            if (marlin_server.print_state == mpsPrinting) {
+                marlin_server.print_state = mpsPausing_Begin;
+            } else {
+                thermalManager.setTargetHotend(0, 0);
+            }
+            nozzle_fan_error = 0;
+            error_dialog_show = true;
+            _dbg("nozzle fan not spin dialog show!");
+            osDelay(50);
+        } else if (print_fan_error == 5) {
+            set_warning(WarningType::PrintFanError);
+            if (marlin_server.print_state == mpsPrinting) {
+                marlin_server.print_state = mpsPausing_Begin;
+            }
+            print_fan_error = 0;
+            error_dialog_show = true;
+            _dbg("print fan not spin dialog show!");
+            osDelay(50);
+        }
+    }
```

Comment:
As mentioned above, this is just too much of logic for marlin_server. I believe it shouldn't be here. It fits the best to Marlin itself, but we don't have fan monitoring there, so the second-best option seems to create a separate module in our firmware.

Reply:
Solved